src/axolotl/axolotl.rs

```diff
@@ -18,50 +18,80 @@ pub trait Axolotl {
 
 
     fn derive_initial_root_key_and_chain_key(
-        local_identity_remote_handshake_dh_secret : &DHShared<Self::IdentityKey>, 
-        local_handshake_remote_identity_dh_secred : &DHShared<Self::IdentityKey>, 
-        local_handshake_remote_handshake_dh_secret : &DHShared<Self::IdentityKey>) 
+        &self,
+        local_identity_remote_handshake_dh_secret : &Self::SharedSecret, 
+        local_handshake_remote_identity_dh_secred : &Self::SharedSecret, 
+        local_handshake_remote_handshake_dh_secret : &Self::SharedSecret) 
     -> (Self::RootKey, Self::ChainKey);
 
     // This is the DH future secrecy ratchet/
     fn derive_next_root_key_and_chain_key(
+        &self,
         root_key : Self::RootKey, 
-        ratchet : &<Self::RatchetKey as DH>::Shared) 
+        ratchet : &Self::SharedSecret) 
     -> (Self::RootKey, Self::ChainKey);
 
     //This is the SCIMP style forward secrecy chain key iteration.
     fn derive_next_chain_and_message_key(
+        &self,
         chain_key : &Self::ChainKey) 
     -> (Self::ChainKey, Self::MessageKey);
 
     fn encrypt_message(
+        &self,
         message_key : &Self::MessageKey,
         plaintext : &Self::PlainText) 
     -> Self::CipherText;
 
     fn decrypt_message(
+        &self,
         message_key : &Self::MessageKey,
         cyphertext : &Self::CipherText) 
     -> Option<Self::PlainText>;
 
     fn authenticate_message(
+        &self,
         message : &AxolotlMessage<Self>, 
         message_key : &Self::MessageKey, 
-        sender_identity : &DHPublic<Self::IdentityKey>, 
-        receiver_identity : &DHPublic<Self::IdentityKey>)
+        sender_identity : &Self::PublicKey, 
+        receiver_identity : &Self::PublicKey)
     -> Self::Mac;
 
     fn ratchet_keys_are_equal(
-        key0 : &DHPublic<Self::RatchetKey>, 
-        key1 : &DHPublic<Self::RatchetKey>) 
+        &self,
+        key0 : &Self::PublicKey, 
+        key1 : &Self::PublicKey) 
     -> bool;
 
-    fn generate_ratchet_key_pair() 
-    -> DHKeyPair<Self::RatchetKey>;
+    fn generate_ratchet_key_pair(&self) -> KeyPair<Self>;
 
-    fn future_message_limit() -> u32;
+    fn derive_shared_secret(&self, key : &Self::PrivateKey, public : &Self::PublicKey) -> Self::SharedSecret;
 
-    fn chain_message_limit() -> u32;
+    fn derive_public_key(&self, key : &Self::PrivateKey) -> Self::PublicKey;
 
-    fn skipped_chain_limit() -> usize;
+    fn future_message_limit(&self) -> u32;
+
+    fn chain_message_limit(&self) -> u32;
+
+    fn skipped_chain_limit(&self) -> usize;
+}
+
+
+pub struct KeyPair<T> where T:Axolotl {
+    pub key : T::PrivateKey,
+    pub public : T::PublicKey,
+}
+
+pub struct ExchangedPair<T> where T:Axolotl {
```

Comment:
Exchanged Pair has always seemed questionable to me if it adds any value. Coupling local and remote Pub/Privates into key pairs is really unintuitive to me, but maybe i'm the only one